Добавлен обработчик ошибок 403 и 404

toyrik opened the pull request:
Внесены изменения в классе Router добавлен метод getControllerFIleName для поиска пути до файла
контроллера, в метод callControllerAction добавлена генерация ощибки в случае если файл контроллера не найден.
В классах SmvcAccesException и SmvcRoutingException изменены коды
генерируемых ошибок

toyrik committed:
Добавлен обработчик ошибок 403 и 404

Внесены изменения в классе Router добавлен метод getControllerFIleName для поиска пути до файла
контроллера, в метод callControllerAction добавлена генерация ощибки в случае если файл контроллера не найден.
В классах SmvcAccesException и SmvcRoutingException изменены коды
генерируемых ошибок

toyrik committed:
Переносим перехват исключений в метод `Application::run()`

Добавлен интерфейс для классов "обработчиков исключений".
В базовом "перехватчике" добавлена переадресация исключений
пользовательскому обработчику.

vedro-compota commented on src/ExceptionHandler.php:
```diff
@@ -0,0 +1,36 @@
+    public function handleException(\Exception $exception): void
+    {
+        $handlers = Config::get('core.handlers');
+        foreach ($handlers as $exceptionName => $handlerName){
```
in_array() не поможет?

vedro-compota commented on src/interfaces/ExceptionsHandler.php:
```diff
@@ -0,0 +1,14 @@
+interface ExceptionsHandler
+{
+    /**
+     * 
```
комментирии плиз

vedro-compota commented on src/Router.php:
```diff
@@ -49,12 +49,17 @@ protected function __construct() {}
      * @return $this
```
откатить правки, или объяснить зачем нужно

toyrik replied:
Если откатить правки то ошибка генерируется в другом месте и не попадает в обработчик

vedro-compota replied:
ок

vedro-compota commented on src/ExceptionHandler.php:
```diff
@@ -0,0 +1,36 @@
+                $thatHandler = Config::getObject('core.handlers.'.$exceptionName);
+                $thatHandler->handleException($exception);
+            }else{
+                $this->displayException($exception);
```
лучше бросить исключение без обработчика ещё раз

toyrik replied:
Поправил

vedro-compota reviewed (changes requested):
нужны ещё правки, написал комменты

toyrik committed:
Добавляем постфикс в название интерфейса обработчика ошибок

В классе ExceptionHandler убран метод displayExeption, теперь если
для исключения не найдено обработчика - выбрасывается тоже самое исключение.
Из класса Router убрана проверка на наличие файла.

toyrik committed:
Рефакторим метод обработки ошибок

В классе ротера в методе вызова действия контроллера добавлена проверка
на наличие файла.

vedro-compota commented on src/ExceptionHandler.php:
```diff
@@ -0,0 +1,35 @@
+use ItForFree\SimpleMVC\Config;
+use ItForFree\SimpleMVC\interfaces\ExceptionHandlerInterface;
+
+class ExceptionHandler implements ExceptionHandlerInterface
```
менеджер исключений не должен имплементировать этот интерфейс, так как этот интрфейс создан, чтобы быть шаблоном для пользовательских обработчиков.

vedro-compota commented on src/ExceptionHandler.php:
```diff
@@ -0,0 +1,35 @@
+        if(array_key_exists(get_class($exception), $handlers)){
+            $exceptionName = get_class($exception);
+            $thatHandler = Config::getObject('core.handlers.'.$exceptionName);
+            $thatHandler->handleException($exception);
```
перед вызовом метода проверить, что объект реализует интерфейс, а иначе бросить исключение нужного типа

toyrik committed:
Добавляем проверку реализации интерфейса пользовательским обработчиком

toyrik commented:
Поправил
